oak/server/rust/oak_runtime/src/node/grpc_server.rs

```diff
@@ -0,0 +1,255 @@
+}
+
+/// Clone implementation without `thread_handle` copying to pass the node to other threads.
+impl Clone for GrpcServerNode {
+    fn clone(&self) -> Self {
+        Self::new(
+            &self.config_name,
+            self.runtime.clone(),
+            self.writer,
+            self.address.clone(),
+        )
+    }
+}
+
+impl Display for GrpcServerNode {
+    fn fmt(&self, f: &mut Formatter) -> Result<(), fmt::Error> {
+        write!(f, "GrpcServerNode({})", self.config_name)
+    }
+}
+
+impl GrpcServerNode {
+    /// Creates a new [`GrpcServerNode`] instance, but does not start it.
+    pub fn new(
+        config_name: &str,
+        runtime: RuntimeProxy,
+        writer: Handle,
+        address: SocketAddr,
+    ) -> Self {
+        Self {
+            config_name: config_name.to_string(),
+            runtime,
+            writer,
+            thread_handle: None,
+            address: address,
+        }
+    }
+
+    /// Processes an incoming `http_request`, decodes a gRPC request
+    async fn serve(
+        &self,
+        http_request: hyper::Request<hyper::Body>,
+    ) -> Result<hyper::Response<hyper::Body>, hyper::Error> {
+        // Parse HTTP header.
+        let grpc_method = http_request.uri().path();
+
+        // Aggregate the data buffers from an HTTP body asynchronously.
+        let http_body = hyper::body::aggregate(http_request)
+            .await
+            .map_err(|error| {
+                warn!("Couldn't aggregate request body: {}", error);
+                error
+            })?;
+
+        // Create a gRPC request from an HTTP body.
+        Self::decode_grpc_request(grpc_method, &http_body)
+            .and_then(|request| {
+                // Process a gRPC request and send it into the Runtime.
+                self.process_request(request)
+                    .and_then(|response_reader| {
+                        // Read a gRPC response from the Runtime.
+                        self.process_response(response_reader)
+                            .and_then(|body| {
+                                // Send gRPC response back to the HTTP client.
+                                Ok(Self::http_response(http::StatusCode::OK, body))
+                            })
+                            .or_else(|error| {
+                                error!("Couldn't process gRPC response: {}", error);
+                                Ok(Self::http_response(
+                                    http::StatusCode::INTERNAL_SERVER_ERROR,
+                                    vec![],
+                                ))
+                            })
+                    })
+                    .or_else(|error| {
+                        error!("Couldn't process gRPC request: {}", error);
+                        Ok(Self::http_response(
+                            http::StatusCode::INTERNAL_SERVER_ERROR,
+                            vec![],
+                        ))
+                    })
+            })
```

Comment:
Could this be encapsulated by a `Result<hyper::Response<hyper::Body>, Error>`, where `Error` is a newly defined enum, and just convert `Error` to the appropriate HTTP response here? Then you can use `?` throughout.

Reply:
Thanks, now it looks much nicer!
Done